cro-clib/src/address.rs

```diff
@@ -0,0 +1,84 @@
+use crate::types::{CroAddressPtr, CroResult};
+use std::ffi::CString;
+use std::ptr;
+/// print address information
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_print_address(
+    address_ptr: CroAddressPtr,
+    buf: *mut u8,
+    buf_length: u32,
+) -> CroResult {
+    let address = address_ptr.as_mut().expect("get address");
+    let address_string = address.address.to_string();
+    if address_string.as_bytes().len() >= buf_length as usize {
+        return CroResult::fail();
+    }
+    ptr::write_bytes(buf, 0, buf_length as usize);
+    ptr::copy_nonoverlapping(
+        address_string.as_bytes().as_ptr(),
+        buf,
+        address_string.as_bytes().len(),
+    );
+    CroResult::success()
+}
+
+/// print address information
+/// 32 bytes
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_export_private(address_ptr: CroAddressPtr, dst: *mut u8) -> CroResult {
+    let address = address_ptr.as_mut().expect("get address");
+    let src = address.privatekey.serialize();
+    assert!(32 == src.len());
+    ptr::copy_nonoverlapping(src.as_ptr(), dst, src.len());
+    CroResult::success()
+}
+
+/// print address information
+/// minimum byte length 100 is necessary
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_get_printed_address(
+    address_ptr: CroAddressPtr,
+    address_output: *mut u8,
+    address_output_length: u32,
+) -> CroResult {
+    let address = address_ptr.as_mut().expect("get address");
+    let src_string = CString::new(address.address.as_bytes()).expect("get cstring");
+    let src = src_string.to_bytes_with_nul();
+    if src.len() > address_output_length as usize {
+        return CroResult::fail();
+    }
+    ptr::copy_nonoverlapping(src.as_ptr(), address_output, src.len());
+    CroResult::success()
+}
+
+/// print address information
+/// minimum 32 length is necessary
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_get_raw_address(
+    address_ptr: CroAddressPtr,
+    address_output: *mut u8,
+    address_output_length: *mut u32,
+) -> CroResult {
+    if address_output.is_null() {
+        return CroResult::fail();
+    }
```

Comment:
perhaps not "print", but "extract" (as this is raw, not printable / textual format)?

Reply:
ok